Allow dominated planets to be freed

tiennou opened the pull request:
First PR, please be gentle ! This makes it possible to give back planets after you dominated them, just because I dominated a few pirate worlds while doing the FW storyline, namely Gienah, Algenib & Almach, but as the first two are switched to Federation after the last battle I felt bad ;-).

toilethinges commented:
Nice! Plenty of people have requested this, because they feel bad about conquering, for example, Smuggler's Den, or Markab.

LocalGod79 commented:
could you perhaps add a reputation boost with the government associated with the freed planet?  not enough to compensate for the defence fleet, but...something?

tiennou commented:
I had actually added a +10 reputation boost, but I removed it since I'm just beginning to tinker with the code and completely unaware of how domination works — since I only dominated some pirate systems, I'm still not sure of what kind of value is acceptable here... Maybe trigger a new TRIBUTED ShipEvent, as well as "tributed"/"relinquished" penalties on governments ? Also (if I can find a nice way to do it) I would like to make the messages customizable by government too.

LocalGod79 commented:
i'm thinking rather than a flat rate, a number related to the required tribute or tribute threshold for that planet (something easily proportioned to the number of people you killed in the defence fleet) - perhaps one of those numbers divided by 100?

while destroying the defence fleet of a planet, you are damaging your reputation with that government by one point for each crew on the ships you're disabling/destroying (and others, by varying amounts defined by governments.txt).

like i said - if your reputation was 'more or less neutral' when you demanded tribute, went very negative while destroying the defence fleet - your reputation should still be lower than 'more or less neutral' after you return their freedom...

jmhorjus commented:
Style guide note:
You should tab-ify your new sections of C++ code (i.e. "[space][space][space][space][some code]" should be "[tab][some code]).

Other than that, this looks fine to me (haven't tested personally yet though). Maybe replace the hitch-hiker's reference with a more generic line? I haven't seen any other random jokes or references hiding out in the code base, so it might stand out. :)

egilburg commented:
Perhaps also make it so planets automatically relinquish when their allegiance changes?

tiennou commented:
Fixed the space-indent by manually amending, and rebased against `master`.

@LocalGod79 I feel that depending on the defending fleet's crew count is too hard to do — right now it's not tracked anywhere and I don't see an easy way to keep track of the defender "kill count" to do that. I can do something like `tribute / 100` easily, or maybe something more clever like keeping track of the date the planet surrendered, the government's reputation at the time of the surrender and do some math so that the "relinquish reputation" is 1) always smaller that the reputation we had and 2) it decreases the more you keep tributing the planet. What do you think ?

@jmhorjus Yeah, sorry for the cheesy reference :wink:. I didn't have any nice idea, and (as I said earlier) I felt constrained by the one-size-fits-all of the current code. If you guys have anything better, I'll gladly take it.

@egilburg I happened to notice that my initial example (namely Gienah & Algenib) are only back under ~~Federation~~Republic control for 50 days — as such, I don't think I would have liked it :wink:. Thinking about it, I think that's a point for having TRIBUTE/RELINQUISH ShipEvents — we'd be able to switch governments and automatically offend them because we tribute one of their planets, or automatically relinquish in case we're friends with them.

egilburg commented:
@tiennou I came from the perspective that if you dominate a planet belonging to a centralized government, it should incur a major penalty (perhaps an automated state of war) with all systems of that government, and freeing any planets you conquered should be a condition of becoming non-hostile. Of course it goes beyond scope of PR, it just doesn't make sense how you can dominate a random planet belonging to Federation, for example, and rest of federation goes like "ok fine we can still be friends". So if you start conquering a major faction's planets, it's an all out war until you either conquer the entire faction or you release all the planets.

It would be different for non-aligned or loosely-aligned factions like Pirates - they don't have centralized government so you can capure an individual pirate planet and rest won't care much about it.

tiennou commented:
@egilburg I'm in complete agreement with what you say. I was actually quite surprised to see those planets switch government *without* triggering an all-out war with the Republic, and then I found out that you can't relinquish them — hence this PR.

lehgogh commented:
@egilburg However, if you capture several planets or a major planet, *that* would start a war against the Pirates.

LocalGod79 commented:
@jmhorjus there are various "easter egg" references made by the author - i don't really see any problem with adding some flavour like this, as long as the context doesn't depend on the user knowing the reference.

@tiennou i understand the difficulty in finding the reputation loss - fleet/ship spawning is random, and so is their crew compliment - it's why the first paragraph suggested using either the tribute or threshold as a correction base (as they will be correlated to the fleet that's sent after you when you demand tribute).  the rest was just to inform you of the (to be ignored) dynamic (and the difficulties lying therein).

for most intents and purposes, if you demand tribute from a planet, you will kill enough ships to drive your reputation with that government into a state where you are at war with the whole faction.  i always thought that if it was a minor world, and you'd already killed enough pirates, the faction's government might be lubed enough to let it slide...

that said, i feel like many players that get to the point where they want to/could contribute to E.S., forget what it was like to play the intended game.  i.e. if you've killed so many korath that you could dominate all of the syndicate, and their fleets are still friendly - that was not an intended result.  it came from running out of content, and grinding out more.

tiennou commented:
@endless-sky Can I respectfully ask for some architectural insight on this ?

- I don't see why the tribute gets stored in `PlayerInfo` (instead of a more helpful `domination_date`) — apart from displaying the list of tributes in the UI. Since the tribute value is (now) accessible from `Planet` that could be done.
- I've been thinking that having the attribute set on `Planet` instead would allow events to use that (and thus be used to have missions from your own private empire).
- Any opinion on how to handle the reputation change ? I was thinking of keeping track of the domination date, the reputation at the time of the domination, and doing some math with that, but maybe you can think of something else ?
- Do you like it :stuck_out_tongue_winking_eye: ?

LocalGod79 commented:
@tiennou - the entire tribute system is wildly undeveloped as a whole because it was never intended to be played as intensely as it is.  with that in mind, i believe the real question is: would a more developed tribute system be a worthy effort?

tiennou commented:
Oh right, seeing it's missing the ability to relinquish planets, I can get that ;-), I was just aiming at feature-parity (because of EVN mostly, but also "symmetry"). If developing the tribute system is considered worthy, I'll gladly do it, if only for the sake of "content".

LocalGod79 commented:
personally, i believe M.Z. would prefer that you addressed other feature requests/issues (that he's marked as milestones) rather than embark on something like this.

i look at other reasonable code pulls that have been made, that now also contain conflicts as code has evolved around them - one assumes: because there are bigger fish to fry.

that said, what you do with your own time is yours - i'm just hoping to help M.Z. direct some traffic.

endless-sky commented:
This seems like a really complicated PR, just to make it possible to relinquish control of a planet. All that's necessary to do that is to:

* erase the "tribute: <name>" condition in PlayerInfo.
* remove the planet from dominatedPlanets in Politics.

So, all the work could be done in HailPanel, if Politics::DominatePlanet() were changed to take a second optional argument that signals whether you're dominating or releasing the planet. Then you could just do this when the button is pressed:

```C++
if(GameData::GetPolitics().HasDominated(planet))
{
    GameData::GetPolitics().DominatePlanet(planet, false);
    player.Conditions().erase("tribute: " + planet->Name());
    message = "Thank you for granting us our freedom!"
}
else
    message = planet->DemandTribute(player);
```

tiennou commented:
Sorry about it being complicated, I was a little misled by trying to handle reputation changes, and got carried away. I'll rewrite it ASAP.

tiennou commented:
Rebased.

endless-sky commented:
Maybe have the button say "Relinquish _Tribute" instead of "_Relinquish Planet"? That makes it clearer that you're giving up the tribute, and it means that the same hotkey (T) works.

Also, try to match the [style](http://endless-sky.github.io/styleguide/styleguide.xml) of the existing code. In particular: curly braces { on their own lines, and no space between "if" and the opening (.

As for the message when you relinquish control, it doesn't make sense to me in context (i.e. in the Hitchhiker books, the humans weren't the overlords of the dolphins, and the "so long" was because the humans were all about to die). Maybe something a bit more generic, like, "Thank you for releasing us from our bondage!"

jmhorjus commented:
Or "Thank you! I guess that's one less pirate we need to send 'protection' money to every day. We'll just delete your name from the database..."
:)

feyone commented:
Point 1: Dominating a world isn't capturing it. It's much more of a "protection racket".
Income from a captured world would be vastly greater / give access to shipyards at half price, etc.
On that basis the world changing allegiance may or may not break the domination.

Point 2 (On relinquishing tribute) - this should have a number of "in game effects"

a) In the FW campaign, there is a sub-thread that the pirate worlds are largely poor and disaffected doing
    what they can to be able to eat regularly, and very few actual pirates. This should be worked into the
    storyline where offering to relinquish tribute strengthens the council decision to offer pirates
    independence in return for abandoning piracy / boosts your rep with the FW/Navy.

b) Boosts rep with that world and makes them friendly
    10% discount on all purchases / +5% paid on all trades/sales, etc.
     - Worth more than the tribute, but you have to be HERE to collect on it.

KiLEdEnNis commented:
1: what exactly is your point here?

2a: we have been told tributes are more of a late game extra income for players than something to actually use in a story

2b: once ya receive tribute from a planet they're already friendly to ya, apart from that, a planet is part of a government, and capturing a pirate planet won't make pirates less hostile
also, discounts & extra money on tributes would be OP to me: you already get free money everyday
